File: src/es/um/josecefe/rueda/resolutor/ContextoResolucionSimple.java
/*
 * Copyright (c) 2016-2017. Jose Ceferino Ortega Carretero
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package es.um.josecefe.rueda.resolutor;

import es.um.josecefe.rueda.modelo.*;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.*;

/**
 * @author josecefe
 */
class ContextoResolucionSimple {
    final List<Dia> dias;
    //Map<Dia, Integer> diasIndex;
    final List<Participante> participantes;
    //Map<Participante, Integer> participantesIndex;
    final Map<Dia, List<AsignacionDiaSimple>> solucionesCandidatas;
    final HashMap<Dia,HashMap<Set<Participante>,AsignacionDiaSimple>> mapaParticipantesSoluciones;
    final Map<Participante, Double> coefConduccion; //Que tanto por 1 supone que use el coche cada conductor

    ContextoResolucionSimple(final Set<Horario> horarios) {
        //Construimos la lista de Dias (usamos un lista para tener un orden)
        Set<Dia> diasSet = new TreeSet<>(); //Lo queremos ordenado
        for (Horario horario : horarios) {
            diasSet.add(horario.getDia());
        }
        dias = new ArrayList<>(diasSet); // Podría ser un array si lo vemos necesario

        //Construimos ahora la lista de conductores (también usamos una lista por lo mismo que en dias
        Set<Participante> participanteSet = new TreeSet<>(); //Lo queremos ordenado
        for (Horario horario : horarios) {
            Participante participante = horario.getParticipante();
            participanteSet.add(participante);
        }
        participantes = new ArrayList<>(participanteSet);

        // Y aquí ajustamos el coeficiente que nos mide su grado de participación (gente que solo va parte de los días)
        Map<Participante, Long> vecesParticipa = horarios.stream().collect(groupingBy(Horario::getParticipante, counting()));
        long maxVecesParticipa = vecesParticipa.values().stream().mapToLong(Long::longValue).max().orElse(0L);
        coefConduccion = new LinkedHashMap<>(participantes.size());
        for (Participante p : participantes) {
            coefConduccion.put(p, (double) maxVecesParticipa / vecesParticipa.get(p) - 0.001); // Restamos 0.001 para evitar que al redondear ciertos casos se desmadren
        }

        // Vamos a trabajar día a día
        //Para parelizar: solucionesCandidatas = dias.stream().collect(toMap(ind -> ind, d -> {
        solucionesCandidatas = new LinkedHashMap<>();
        mapaParticipantesSoluciones = new HashMap<>();
        for (Dia d : dias) {
            ArrayList<AsignacionDiaSimple> solucionesDia = new ArrayList<>();
            HashMap<Set<Participante>,AsignacionDiaSimple> mapaParticipantesAsignacionDia = new HashMap<>();
            Set<Horario> horariosDia = horarios.stream().filter(h -> h.getDia() == d).collect(toSet());
            Map<Participante, Horario> participanteHorario = horariosDia.stream().collect(toMap(Horario::getParticipante, Function.identity()));
            List<Participante> participantesDia = horariosDia.stream().map(Horario::getParticipante).sorted().collect(toList());
            final int nParticipantesDia = participantesDia.size();

            // Para cada hora de entrada, obtenemos los conductores disponibles
            Map<Integer, Set<Participante>> entradaConductor = horariosDia.stream().filter(Horario::isCoche)
                    .collect(Collectors.groupingBy(Horario::getEntrada, Collectors.mapping(Horario::getParticipante, toSet())));
            // Para comprobar, vemos los participantes, sus entradas y salidas
            Map<Integer, Long> nParticipantesIda = horariosDia.stream()
                    .collect(Collectors.groupingBy(Horario::getEntrada, Collectors.mapping(Horario::getParticipante, Collectors.counting())));
            Map<Integer, Long> nParticipantesVuelta = horariosDia.stream()
                    .collect(Collectors.groupingBy(Horario::getSalida, Collectors.mapping(Horario::getParticipante, Collectors.counting())));
            // Generamos todas las posibilidades y a ver cuales sirven...
            List<Iterable<Set<Participante>>> conductoresDia = entradaConductor.keySet().stream()
                    .map(key -> new SubSets<>(entradaConductor.get(key), 1, entradaConductor.get(key).size())).collect(Collectors.toList());
            Combinador<Set<Participante>> combinarConductoresDia = new Combinador<>(conductoresDia);

            for (List<Set<Participante>> condDia : combinarConductoresDia) {
                final Set<Participante> selCond = condDia.stream().flatMap(Collection::stream).collect(toSet());
                // Validando que hay plazas suficientes sin tener en cuenta puntos de encuentro
                Map<Integer, Integer> plazasIda = selCond.stream()
                        .map(participanteHorario::get).collect(
                                Collectors.groupingBy(Horario::getEntrada, Collectors.summingInt(h -> h.getParticipante().getPlazasCoche())));
                Map<Integer, Integer> plazasVuelta = selCond.stream()
                        .map(participanteHorario::get)
                        .collect(Collectors.groupingBy(Horario::getSalida, Collectors.summingInt(h -> h.getParticipante().getPlazasCoche())));

                if (nParticipantesIda.entrySet().stream().allMatch(e -> plazasIda.getOrDefault(e.getKey(), 0) >= e.getValue())
                        && nParticipantesVuelta.entrySet().stream().allMatch(e -> plazasVuelta.getOrDefault(e.getKey(), 0) >= e.getValue())) {

                    // Obtenemos la lista de posibles lugares teniendo en cuenta quien es el conductor
                    List<Iterable<Lugar>> posiblesLugares = participantesDia.stream().map(Participante::getPuntosEncuentro).
                            collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
                    posiblesLugares.addAll(selCond.stream().sorted().map(Participante::getPuntosEncuentro).
                            collect(ArrayList::new, ArrayList::add, ArrayList::addAll));

                    int mejorCoste = Integer.MAX_VALUE;
                    Map<Participante, Lugar> mejorLugaresIda = null, mejorLugaresVuelta = null;

                    for (List<Lugar> selLugares : new Combinador<>(posiblesLugares)) {
                        Map<Participante, Lugar> lugaresIda, lugaresVuelta;
                        lugaresIda = new HashMap<>();
                        lugaresVuelta = new HashMap<>();
                        Iterator<Lugar> il = selLugares.subList(nParticipantesDia, selLugares.size()).iterator();
                        for (int i = 0; i < nParticipantesDia; i++) {
                            lugaresIda.put(participantesDia.get(i), selLugares.get(i));
                            lugaresVuelta.put(participantesDia.get(i), selCond.contains(participantesDia.get(i))
                                    ? il.next() : selLugares.get(i));
                        }
                        Map<Integer, Map<Lugar, Integer>> plazasDisponiblesIda = selCond.stream()
                                .collect(Collectors.groupingBy(p -> participanteHorario.get(p).getEntrada(), Collectors.groupingBy(lugaresIda::get, Collectors.summingInt(Participante::getPlazasCoche))));

                        Map<Integer, Map<Lugar, Integer>> plazasDisponiblesVuelta = selCond.stream()
                                .collect(Collectors.groupingBy(p -> participanteHorario.get(p).getSalida(), Collectors.groupingBy(lugaresVuelta::get, Collectors.summingInt(Participante::getPlazasCoche))));
                        // Para comprobar, vemos los participantes, sus entradas y salidas
                        Map<Integer, Map<Lugar, Long>> plazasNecesariasIda = horariosDia.stream()
                                .collect(Collectors.groupingBy(Horario::getEntrada, Collectors.groupingBy(h -> lugaresIda.get(h.getParticipante()), Collectors.counting())));

                        Map<Integer, Map<Lugar, Long>> plazasNecesariasVuelta = horariosDia.stream()
                                .collect(Collectors.groupingBy(Horario::getSalida, Collectors.groupingBy(h -> lugaresVuelta.get(h.getParticipante()), Collectors.counting())));

                        if (plazasNecesariasIda.entrySet().stream().allMatch(e -> e.getValue().entrySet().stream().allMatch(ll -> ll.getValue() <= plazasDisponiblesIda.get(e.getKey()).getOrDefault(ll.getKey(), 0)))
                                && plazasNecesariasVuelta.entrySet().stream().allMatch(e -> e.getValue().entrySet().stream().allMatch(ll -> ll.getValue() <= plazasDisponiblesVuelta.get(e.getKey()).getOrDefault(ll.getKey(), 0)))) {
                            // Calculamos coste
                            int coste = participantesDia.stream().mapToInt(
                                    p -> (p.getPuntosEncuentro().indexOf(lugaresIda.get(p)) + p.getPuntosEncuentro().indexOf(lugaresVuelta.get(p))) *
                                            (selCond.contains(p) ? Pesos.PESO_LUGAR_CONDUCTOR : Pesos.PESO_LUGAR_PASAJERO)).sum();

                            if (coste < mejorCoste) {
                                mejorCoste = coste;
                                mejorLugaresIda = lugaresIda;
                                mejorLugaresVuelta = lugaresVuelta;
                            }
                        }

                    }
                    if (mejorCoste < Integer.MAX_VALUE) {
                        // Tenemos una solución válida
                        AsignacionDiaSimple asignacionDiaSimple = new AsignacionDiaSimple(selCond, mejorLugaresIda, mejorLugaresVuelta, mejorCoste);
                        solucionesDia.add(asignacionDiaSimple);
                        mapaParticipantesAsignacionDia.put(selCond, asignacionDiaSimple); //Para acelerar la busqueda después
                    }
                }
            }
            solucionesDia.sort(null); //Ordenamos la soluciones parciales tras el barajado
            solucionesCandidatas.put(d, solucionesDia);
            mapaParticipantesSoluciones.put(d, mapaParticipantesAsignacionDia);
        }
    }
}

File: src/es/um/josecefe/rueda/resolutor/ContextoResolucionSimple.kt
/*
 * Copyright (c) 2016-2017. Jose Ceferino Ortega Carretero
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package es.um.josecefe.rueda.resolutor

import es.um.josecefe.rueda.modelo.*
import es.um.josecefe.rueda.util.Combinador
import es.um.josecefe.rueda.util.SubSets
import es.um.josecefe.rueda.util.combinations
import java.util.*
import kotlin.math.max

/**
 * @author josecefe
 */
internal class ContextoResolucionSimple(horarios: Set<Horario>) {
    val dias: List<Dia> = horarios.map { it.dia }.distinct().sorted()
    val participantes: List<Participante> = horarios.map { it.participante }.distinct().sorted()
    val solucionesCandidatas: MutableMap<Dia, List<AsignacionDiaSimple>>
    val mapaParticipantesSoluciones: MutableMap<Dia, Map<Set<Participante>, AsignacionDiaSimple>>
    val coefConduccion: Map<Participante, Double> //Que tanto por 1 supone que use el coche cada conductor
    val diasFijos: Map<Participante, MutableSet<Dia>>
    val minCond: Int
    val mapParticipanteDias: Map<Participante, Set<Dia>>
    val totalPosiblesSoluciones: Long
    private val solucionesNivel: LongArray

    init {
        // Y aquí ajustamos el coeficiente que nos mide su grado de participación (gente que solo va parte de los días)
        val vecesParticipa = horarios.groupingBy { it.participante }.eachCount()
        val maxVecesParticipa = vecesParticipa.values.max() ?: 0
        coefConduccion = participantes.associateWith { maxVecesParticipa.toDouble() / (vecesParticipa[it]?.toDouble() ?: 0.0) - 0.001 }
        // Vamos a trabajar día a día
        //Para parelizar: solucionesCandidatas = dias.stream().collect(toMap(ind -> ind, d -> {
        solucionesCandidatas = LinkedHashMap()
        mapaParticipantesSoluciones = HashMap()
        for (d: Dia in dias) {
            val solucionesDia = mutableListOf<AsignacionDiaSimple>() //Para paralelizar usar Vector en luar de ArrayList
            val mapaParticipantesAsignacionDia = mutableMapOf<Set<Participante>, AsignacionDiaSimple>() //Para paralelizar usar ConcurrentHashMap
            val horariosDia = horarios.filter { it.dia == d }.toSet()
            val participanteHorario = horariosDia.associateBy { it.participante }
            //val participantesDia = horariosDia.map { it.participante }.sortedBy { it.nombre }

            // Para cada hora de entrada, obtenemos los conductores disponibles
            val entradaConductor = horariosDia.filter { it.coche }.groupBy({ it.entrada }, { it.participante })
            // Para comprobar, vemos los participantes, sus entradas y salidas
            val nParticipantesIda = horariosDia.groupingBy { it.entrada }.eachCount()

            val nParticipantesVuelta = horariosDia.groupingBy { it.salida }.eachCount()
            // Generamos todas las posibilidades y a ver cuales sirven...
            val conductoresDia = entradaConductor.keys.map { key ->
                SubSets(entradaConductor.getValue(key), 1, entradaConductor.getValue(key).size).toList()
            }
            val combinarConductoresDia = Combinador(conductoresDia)

            //for (List<Set<Participante>> condDia : combinarConductoresDia) {
            for (condDia in combinarConductoresDia) {
                val selCond = condDia.flatten().toSortedSet()
                // Validando que hay plazas suficientes sin tener en cuenta puntos de encuentro

                val plazasIda = selCond.map { participanteHorario[it] }.groupBy { it!!.entrada }.mapValues { it -> it.value.sumBy { it!!.participante.plazasCoche } }
                val plazasVuelta = selCond.map { participanteHorario[it] }.groupBy { it!!.salida }.mapValues { it -> it.value.sumBy { it!!.participante.plazasCoche } }

                if (nParticipantesIda.entries.all { e -> plazasIda[e.key] ?: 0 >= e.value }
                        && nParticipantesVuelta.entries.all { e -> (plazasVuelta[e.key] ?: 0) >= e.value }) {

                    val asignacionDiaSimple = AsignacionDiaSimple(selCond)
                    solucionesDia.add(asignacionDiaSimple)
                    mapaParticipantesAsignacionDia[selCond] = asignacionDiaSimple //Para acelerar la busqueda después
                }
            }
            solucionesDia.sort()
            solucionesCandidatas[d] = solucionesDia
            mapaParticipantesSoluciones[d] = mapaParticipantesAsignacionDia
        }

        /* Averiguamos los días fijos de cada participantes (aquellos días que forzosamente tiene que llevarse el coche) */
        val diasFijosl: MutableMap<Participante, MutableSet<Dia>> = mutableMapOf()
        for ((dia, solDia) in solucionesCandidatas) {
            val conductores: MutableSet<Participante> = participantes.toMutableSet()
            for (asignacion: AsignacionDia in solDia) conductores.retainAll(asignacion.conductores)
            for (c in conductores) {
                var d = diasFijosl[c]
                if (d == null) {
                    d = HashSet()
                    diasFijosl[c] = d
                }
                d.add(dia)
            }
        }

        diasFijos = diasFijosl.toMap()
        /* Averiguamos cual es el minimo de veces que se debe llevar el coche cada uno a partir del máximo de dias fijos */
        minCond = max(1, diasFijos.values.map { it.size }.max()?.toInt() ?: 0)

        mapParticipanteDias = LinkedHashMap<Participante, Set<Dia>>()
        for (i in participantes.indices) {
            if (participantes[i].plazasCoche > 0) {
                val p: Participante = participantes[i]
                val setDias = horarios.filter { h -> h.coche && h.participante == p }.map { h -> h.dia }.toSet()
                mapParticipanteDias[p] = setDias
            }
        }

        var posiblesSoluciones = 0L
        solucionesNivel = LongArray(dias.size - minCond + 1)

        for (nivel in minCond..dias.size) {
            var comb: Long = 1
            for ((key, value) in mapParticipanteDias) {
                val numVecesCond = max(1,
                        (nivel.toFloat() / coefConduccion.getValue(key) + 0.5f).toInt()) //Para minorar adecuadamente el nivel usando el coeficiente de conductor
                val porDiasFijos = (diasFijos[key]?.size ?: 0).toLong()
                val combinations = combinations(max(1, value.size.toLong() - porDiasFijos),
                        max(0, numVecesCond - porDiasFijos))
                comb *= combinations
            }
            solucionesNivel[nivel - minCond] = comb
            posiblesSoluciones += comb
        }
        totalPosiblesSoluciones = posiblesSoluciones
    }

    fun getSolucionesNivel(nivel: Int) = solucionesNivel[nivel - minCond]
}